### PR TITLE
Param for Max Retrial of Engine Cranking.

### PR DESCRIPTION
**Pull request by loki077**
As we use a common battery for Avionics and Engine Cranking we would like to have an option to limit the number of retrials so that it doesn't drain that battery completely. Obvious choice is to separate the batteries but I still think there could be some use of this feature.

My Implementation is by adding Param MAX_RETRY which If set 0 then there is no limit to retrials. If set to a value greater than 0 then the engine will retry starting the engine this many times before giving up. Also the counter to track retrials resets to zero on disarm (could be done only on power cycle rather than disarm). If needed to be override can be done by setting the param to 0.

I would like to get more ideas on the method of its implementation. I have not yest tested this code.

Other ways to implement this is by just keep increasing the time between retrials.

### CODE REVIEW

**Comment by loki077** [comment]
@peterbarker @tridge FYI

File: libraries/AP_ICEngine/AP_ICEngine.cpp
Comment (robertlong13) [inline libraries/AP_ICEngine/AP_ICEngine.cpp:174]: Underlying type as Int8, not UInt8. Should change this to a max of 127 or should change to uint8

File: libraries/AP_ICEngine/AP_ICEngine.cpp
Comment (robertlong13) [inline libraries/AP_ICEngine/AP_ICEngine.cpp:376]: Should probably also reset the counter when we shut the ignition off. This way, someone can shut the engine off/on to reset if they want to force it to retry again.

File: libraries/AP_ICEngine/AP_ICEngine.cpp
Comment (loki077) [reply libraries/AP_ICEngine/AP_ICEngine.cpp:174]: will fix it.

File: libraries/AP_ICEngine/AP_ICEngine.cpp
Comment (loki077) [reply libraries/AP_ICEngine/AP_ICEngine.cpp:376]: I am yet not sure of the approach of counter being reset is correct way so waiting for few more feedback. But rather then resetting at ignition off maybe we can reset it at Mavlink command of DO_ENGINE_CONTROL.
I will have a look at it.

Comment (peterbarker) [reply libraries/AP_ICEngine/AP_ICEngine.cpp:376]: Agreed with a reset if the user moves away from the auto-restart position and back into it

**Review by peterbarker — COMMENTED** [review]
Needs an autotest

File: libraries/AP_ICEngine/AP_ICEngine.cpp
Comment (robertlong13) [inline libraries/AP_ICEngine/AP_ICEngine.cpp:409]: Anything less than zero should be considered infinite retry. As written, -1 means you can't start the engine ever.

Should be
`if (max_crank_retry > 0 && crank_retry_ct < max_crank_retry) {`

File: libraries/AP_ICEngine/AP_ICEngine.cpp
Comment (robertlong13) [inline libraries/AP_ICEngine/AP_ICEngine.cpp:442]: Whitespace change outside of scope of this PR, but also fixed incorrectly. For our astyle rules, it should actually be:

```
            if (!AP::rpm()->get_rpm(rpm_instance-1, rpm_value) ||
                    rpm_value < rpm_threshold) {
                // engine has stopped when it should be running
                state = ICE_START_DELAY;
                GCS_SEND_TEXT(MAV_SEVERITY_INFO, "Uncommanded engine stop");
            }
  ```
  
  i.e., adding an extra indent to line 438, not 439-441

**Review by robertlong13 — CHANGES_REQUESTED** [review]

File: libraries/AP_ICEngine/AP_ICEngine.cpp
Comment (loki077) [reply libraries/AP_ICEngine/AP_ICEngine.cpp:409]: if (max_crank_retry <= 0 || crank_retry_ct < max_crank_retry) 
   
   this will be a fix as we want it to try infinitely when it is 0 or less, or less then maxcount.

File: libraries/AP_ICEngine/AP_ICEngine.cpp
Comment (loki077) [reply libraries/AP_ICEngine/AP_ICEngine.cpp:442]: gitkraken view messed this up for me.....  thanks for pointing it,,,

File: libraries/AP_ICEngine/AP_ICEngine.cpp
Comment (IamPete1) [inline libraries/AP_ICEngine/AP_ICEngine.cpp:419]: This probably warrant more than a "info" severity level. Probably critical or error. The definitions are here: https://mavlink.io/en/messages/common.html#MAV_SEVERITY

File: libraries/AP_ICEngine/AP_ICEngine.cpp
Comment (IamPete1) [inline libraries/AP_ICEngine/AP_ICEngine.cpp:420]: I think this deserves a comment.

File: libraries/AP_ICEngine/AP_ICEngine.h
Comment (IamPete1) [inline libraries/AP_ICEngine/AP_ICEngine.h:112]: Why not a uint8_t?

File: libraries/AP_ICEngine/AP_ICEngine.cpp
Comment (IamPete1) [inline libraries/AP_ICEngine/AP_ICEngine.cpp:417]: I think we should reset this on a un-commanded stop. You could build up some retrys on initial start then run for some long period of time. Then the engine stops without being commanded and you only have one retry left. Here:
https://github.com/ArduPilot/ardupilot/blob/04ee66bec4f6c5aa8e2f0d21a3facd34f1f8f29b/libraries/AP_ICEngine/AP_ICEngine.cpp#L430

**Review by IamPete1 — CHANGES_REQUESTED** [review]

File: libraries/AP_ICEngine/AP_ICEngine.cpp
Comment (robertlong13) [reply libraries/AP_ICEngine/AP_ICEngine.cpp:417]: This would take a larger refactor of ICE than we wanted to push right now. There's currently no difference in the code between an unsuccessful start attempt and an uncommanded engine stop. Both go through the "engine running" state.

File: libraries/AP_ICEngine/AP_ICEngine.cpp
Comment (robertlong13) [reply libraries/AP_ICEngine/AP_ICEngine.cpp:419]: Tell that to "Uncommanded engine stop" 😆 

I also think this should be critical or error, but I thought I'd get pushback given the severity of the engine stop message

File: libraries/AP_ICEngine/AP_ICEngine.cpp
Comment (robertlong13) [reply libraries/AP_ICEngine/AP_ICEngine.cpp:420]: Great point

File: libraries/AP_ICEngine/AP_ICEngine.h
Comment (robertlong13) [reply libraries/AP_ICEngine/AP_ICEngine.h:112]: Basically just felt bad mixing signed and unsigned for the two, and there's no AP_Uint8 for the param.

File: libraries/AP_ICEngine/AP_ICEngine.cpp
Comment (tridge) [inline libraries/AP_ICEngine/AP_ICEngine.cpp:417]: need to make this reset if we have been running for 5x crank time

**Review by tridge — CHANGES_REQUESTED** [review]

**Review by tridge — APPROVED** [review]

File: libraries/AP_ICEngine/AP_ICEngine.cpp
Comment (peterbarker) [inline libraries/AP_ICEngine/AP_ICEngine.cpp:451]: WARN

File: libraries/AP_ICEngine/AP_ICEngine.cpp
Comment (peterbarker) [reply libraries/AP_ICEngine/AP_ICEngine.cpp:422]: WARN

File: Tools/autotest/quadplane.py
Comment (peterbarker) [inline Tools/autotest/quadplane.py:1128]: `check_context`, similarly elsewhere in here

**Review by peterbarker — CHANGES_REQUESTED** [review]

**Pull request merged**
